feat(app): App MVP follow-up — RBAC, observer-aligned metrics, label … (#1149)
databrickslabs/dqx: docs/dqx/docs/guide/dqx_studio.mdx

---
sidebar_position: 314
---

import Admonition from '@theme/Admonition';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# DQX Studio (No-Code UI)

**DQX Studio** is the web application for authoring, reviewing, running, and monitoring data quality rules without writing code.
It runs as a [Databricks App](https://docs.databricks.com/en/dev-tools/databricks-apps/index.html) inside your workspace and exposes all the core DQX capabilities — profiling, rule authoring, approval workflows, scheduled runs, and results tracking — through a browser.

<Admonition type="tip" title="When to use DQX Studio">
DQX Studio is the recommended entry point for:

> - Data stewards and analysts who want to manage quality rules without code.

Reviewer: ```suggestion
- Data stewards and analysts who want to manage and run quality rules without code-level integration (see [post-factum monitoring](Please also link the Architecture: https://databrickslabs.github.io/dqx/docs/motivation/#post-factum-monitoring))
```

- Teams that need an approval workflow before rules go live.
- Any user who prefers to browse Unity Catalog and author checks through a UI.

For code-level integration into pipelines or notebooks, see the [Programmatic approach](/docs/guide/quality_checks_apply#programmatic-approach).
</Admonition>

## Key concepts

Before opening DQX Studio, it helps to understand a few terms used across the UI.

| Term | Meaning |
| ---- | ------- |

> | **Rule** | A single data quality check — a named condition evaluated against a column, a table, or a SQL expression spanning multiple tables. Rules have a severity (error or warning) and optional labels. |

Reviewer: ```suggestion
| **Rule** | A single data quality check — a named condition evaluated against a single or multiple tables and columns. Rules have a severity (error or warning) and optional labels. |
```

| **Draft** | A rule that has been authored but not yet approved. Drafts are editable and do not run on schedules. |
| **Active rule** | An approved rule. Only active rules are executed by runs and schedules. |
| **Run** | A single execution of one or more active rules against their target tables. Runs produce pass/fail results and optional metrics. |
| **Schedule** | A cadence (e.g. hourly, daily, weekly) that automatically submits runs for a set of rules. |
| **Label** | A free-form tag (such as `pii`, `gold`, `finance`) attached to rules for organisation and filtering. |

## Accessing DQX Studio

Your workspace administrator deploys DQX Studio and shares its URL with you. Once deployed, open it in a browser — authentication uses your Databricks identity, so no separate login is required.

> If DQX Studio isn't available in your workspace yet, ask your admin to follow the [deployment guide](https://github.com/databrickslabs/dqx/blob/v0.13.0/app/DEPLOYMENT.md).

Reviewer: We should move this to installation page

Author: you mean the DEPLOYMENT.md should be moved ti the installation page?

Author: added DQX Studio intallation section

<Admonition type="note" title="What you can see is what you can access">

> DQX Studio browses Unity Catalog using your own identity (on-behalf-of). You will only see catalogs, schemas, tables, and columns that you have been granted access to. Profiler and dry-run jobs read data through a temporary view bound to your permissions, so the studio can never read tables you cannot access yourself.

Reviewer: ```suggestion
DQX Studio browses Unity Catalog using your own identity (on-behalf-of). You will only see catalogs, schemas, tables, and columns that you have been granted access to. Profiling and dry-runs are also executed on your behalf, ensuring the studio can never read data from tables you don’t have permission to access.
```
No need to leak internal implementation

### Single-table rules

Use this flow when your check operates on the columns of one table — for example, "this column must not be null", "this column must be unique", or "values must be within a list".

1. Browse Unity Catalog and pick a target table.
2. Choose a column (or leave it empty for whole-row or custom checks).
3. Pick a check from the catalog of built-in DQX checks (such as `is_not_null`, `is_in_list`, `is_unique`, `regex_match`, `is_in_range`).
4. Fill in any arguments the check needs.
5. Set the severity — **error** (fails the run) or **warning** (reports but doesn't fail).
6. Add labels if you want to group the rule (for example `pii`, `finance`, `critical`).
7. Save the rule as a draft.

For a complete list of built-in checks and their arguments, see [Quality Checks Reference](/docs/reference/quality_checks).

### Cross-table rules

Use this flow when a check needs to compare data across multiple tables — for example, referential integrity ("every `order_id` must exist in `orders`"), reconciliation ("sum of `amount` in `payments` equals total in `invoices`"), or any custom SQL condition.

1. Give the rule a name and description.

> 2. Write a SQL expression that returns rows violating the check (or zero rows when everything is healthy).

Reviewer: ```suggestion
2. Write a SQL query that returns rows violating the check (or zero rows when everything is healthy).
```
sql expression are different than queries. Expressions can only opearate on one table

3. Set severity and labels.
4. Save the rule as a draft.

<Admonition type="tip" title="Cross-table SQL tips">

> - The SQL is evaluated by your workspace — it can reference any tables you have access to.

Reviewer: it reads oddly, workspace by itself does not evaluate anything

- Keep checks focused: one rule per invariant is easier to triage when it fails.
- Prefer fully qualified table names (`catalog.schema.table`) for clarity.
</Admonition>

### Profile and generate

If you're starting from scratch and don't yet know what rules to write, profiling a table is the fastest way to get a draft set.

1. Pick a catalog, schema, and table in the browser.
2. Click **Run profiler**.

> 3. Wait for the job to complete — profiling runs as a serverless Databricks Job and uses your own table permissions, so no elevated access is required.

Reviewer: is it possible to customize it in the app to use non-serverless? if yes this needs to be udpated

Author: no, it is serverless only job

5. Keep the ones you like and discard the rest. Everything you keep is saved as drafts for review.

See [Data Profiling](/docs/guide/data_profiling) for the full list of statistics and candidate check types the profiler produces.

<Admonition type="tip" title="AI-assisted generation">
DQX Studio also supports AI-assisted rule suggestions — you describe the business context in natural language and the studio proposes relevant checks. See [AI-Assisted Generation](/docs/guide/ai_assisted_quality_checks_generation) for the underlying capabilities.
</Admonition>

### Import rules

If you already maintain rules in YAML or JSON, the **Import rules** flow lets you paste them in or upload a file. Imported rules land as drafts so you can review them before they go live.

The accepted formats are the same as the programmatic storage formats — see [Quality Checks Definition](/docs/guide/quality_checks_definition) for the schema.

## Reviewing and approving drafts

Drafts live on the **Drafts & Review** page. The workflow depends on your role:

> - **Authors** see their drafts and can edit, delete, or submit them for approval. They cannot approve their own drafts.

Reviewer: ```suggestion
- **Authors** see their drafts and can edit, delete, or submit them for approval. They cannot approve their own drafts unless they are admins.
```

Author: actually both admin and approver are able to submit author's rules. adapted in the docs as well

- **Approvers and admins** see all drafts across the workspace. They can approve or reject each one with an optional comment.

Approved drafts disappear from this page and show up on **Active Rules**. Rejected drafts remain visible so the author can edit and resubmit.

<Admonition type="note" title="Four-eyes principle">

> By design, an author cannot approve their own draft. This enforces a minimum of two people touching every rule before it becomes active — reducing accidental outages caused by a single contributor.

Reviewer: is this true for admins as well? i guess admin can approve even if he/she is the owner

</Admonition>

## Browsing active rules

The **Active Rules** page lists every approved rule grouped by target table (or under **Cross-table rules** for SQL checks). From here you can:

- **Filter** by catalog, schema, table, or label.
- **Expand** any rule to see its definition, labels, author, approval history, and comments.
- **Download** one or more rules as YAML — useful for version control or sharing with another environment.
- **Delete** a rule (admins and approvers only). Deleted rules are removed from all schedules.

> 

Reviewer: ```suggestion
```

> Rules are the library that runs and schedules draw from — any rule here is eligible to be executed.

Reviewer: I would drop the sentence, it's confusing### 
```suggestion
```

## Running rules

> The **Run Rules** page (visible if you have the Runner privilege) lets you trigger runs manually.

Reviewer: It's also for setting up schedules. I would restructure this part a bit. 
Running rules have 2 options: manually or schedule

1. Pick the rules you want to run — filter by table or labels to select a coherent set.
2. Choose the target tables (if a rule applies to multiple) and any per-run parameters.
3. Click **Run** to submit. DQX Studio creates a Databricks Job run and shows you a progress indicator.

You can leave the page while a run is in flight — it will continue in the background and surface in **Runs History** when it finishes.

> ### Dry-run before committing

Reviewer: Move it to the section about authoring rules, it's more relevant there

Each rule supports a **dry run** that evaluates the check against a small sample of the target table without persisting results. Use it while authoring to sanity-check that your rule behaves as expected.

## Schedules

> Runners can put any set of active rules on a schedule so they execute automatically.

Reviewer: ```suggestion
Users with `Runner` privilege can put any set of active rules on a schedule so they execute automatically.
```

4. Save. DQX Studio now triggers runs automatically on that cadence.

Schedules survive restarts and deduplicate — if a scheduled run is still in flight when the next tick arrives, the studio will not double-submit.

<Admonition type="tip" title="Timezones">
All schedule times are interpreted in the timezone configured on the **Configuration** page. Ask your admin to set this to your team's primary timezone before creating schedules.
</Admonition>

### Pausing and editing schedules

Any schedule can be paused (without losing its configuration) and resumed later. Editing a schedule immediately reschedules the next run — there is no need to delete and recreate.

## Runs history

The **Runs History** page is read-only and visible to everyone. It shows every run DQX Studio has executed, sorted by most recent first.

For each run you can see:

- When it started and how long it took.
- Whether it was triggered manually or by a schedule (and which schedule).
- Which rules it covered.
- Pass, warning, and error counts per rule.
- A link to drill into individual row-level failures.

> Use this view to spot regressions, audit rule behaviour over time, and share results with stakeholders.

Reviewer: Link also documention for DQX dashboard. It's another way to monitoring execution

## Configuration (admin only)

> The **Configuration** page is where admins manage workspace-level settings that affect all DQX Studio users.

Reviewer: ```suggestion
The **Configuration** page is where admins manage DQX Studio-level settings that affect all DQX Studio users.
```

| Setting | What it controls |
| ------- | ---------------- |
| **Timezone** | The timezone used to interpret and display all schedule times and run timestamps. |

> | **Labels** | The catalog of labels users can attach to rules. Defining labels centrally keeps tagging consistent. |

Reviewer: ```suggestion
| **Labels** | Labels users can attach to rules. Defining labels centrally keeps tagging consistent. |
```
